### DISCUSSION TITLE
Report File Size Limit when Running PMD?

### DISCUSSION THREAD

**Discussion by cyberjj999** [pmd/pmd#3022] [Q&A]
Hi everyone, I have a simple question to ask - all I wanted to do is to limit the report size when running PMD against my ruleset and java files. 
Based on my experience, when running PMD with my file of interest, I have always gotten up to 70mb (and rising) size of the report, I believe that the size is too large for the number of java files that I am running PMD through. 

I was wondering if I can set a limit of the report and make PMD stop executing when the report reaches a certain size? i.e. 50mb? 

Thank you very much for taking the time to read this 👍

**Comment by oowekyala** [comment]
Hi, what report format are you using, and roughly how many files are you processing?

Currently I don't think there is a way to stop the run. But it might make sense to cap the number of reported violations.

**Comment by cyberjj999** [comment]
@oowekyala Thanks for the quick response!
I am using report format of xml, but im definitely flexible with other types. As for the number of files, I don't have an exact number, what I'm doing is throwing in a folder that contains the source code of some of the projects that I have, so the number isn't definite.

As for capping the number of reported violations, could you give some example? Much thanks!

**Comment by oowekyala** [comment]
> As for capping the number of reported violations, could you give some example? Much thanks!

This is not possible right now but might make sense as a new feature.

If you're running all rules on an unknowly big number of file we can't really say if it's normal that you have many violations. Maybe you could try on a smaller sample of files and check that the report is correct.

**Comment by cyberjj999** [comment]
I see. I am positive that the report is correct, just that theres a shit ton of data haha, was thinking if it can be tidier etc. But I appreciate your help mate!
Let me try to explore further, I'll keep this open in case anyone has some ideas haha

**Comment by adangel** [comment]
Depending on what you want to achieve:
* If you want to introduce PMD (or any other SCA tool) to existing code base, please read https://pmd.github.io/latest/pmd_userdocs_best_practices.html - Basically you start with a small set of rules and work from there. As otherwise - as you have seen yourself - you might be overwhelmed by the found violations
* If you want to get an overview of the quality of the projects, then you can consider running PMD several times on a smaller file set (e.g. do not analyze all projects in one run, but project by project). However, the question is: How useful would be the result? The number of violations don't tell you, how critical they are. And violations might be tolerated in code that's not changing (when changing bad code you might introduce bugs, but if the bad code is correct but not changed, there is no immediate issue).

Oh, and not to forget: If you enable all rules, you might never be able to fix all violations, because rules could contradict each other. You basically always need to choose the rules that are right for you.
And some rules can be configured - and the default parameters might not be what you need in your project.

**Comment by linusjf** [comment]
> Depending on what you want to achieve:
> 
> * If you want to introduce PMD (or any other SCA tool) to existing code base, please read https://pmd.github.io/latest/pmd_userdocs_best_practices.html - Basically you start with a small set of rules and work from there. As otherwise - as you have seen yourself - you might be overwhelmed by the found violations
> * If you want to get an overview of the quality of the projects, then you can consider running PMD several times on a smaller file set (e.g. do not analyze all projects in one run, but project by project). However, the question is: How useful would be the result? The number of violations don't tell you, how critical they are. And violations might be tolerated in code that's not changing (when changing bad code you might introduce bugs, but if the bad code is correct but not changed, there is no immediate issue).
> 
> Oh, and not to forget: If you enable all rules, you might never be able to fix all violations, because rules could contradict each other. You basically always need to choose the rules that are right for you.
> And some rules can be configured - and the default parameters might not be what you need in your project.

If there really is a concern about the file size, PMD can consider breaking up the report into multiple output files based on some criteria such as priority etc. This can also help in deciding which violations to tackle first.

**Comment by linusjf** [comment]
> > As for capping the number of reported violations, could you give some example? Much thanks!
> 
> This is not possible right now but might make sense as a new feature.
> 
> If you're running all rules on an unknowly big number of file we can't really say if it's normal that you have many violations. Maybe you could try on a smaller sample of files and check that the report is correct.

Any particular scenarios when the size of the file would be an issue:

1) Obviously, reading it through a text editor can be a problem.
2) Another would be loading such a large file in the browser.

One option would be to use some sort of file splitting mechanism to break up the file into smaller ones. Is there one for xml files? Or should PMD do this itself based on a file size criteria?
